### ISSUE TITLE
vSphere Volume Attachment Fails with Error "unknown SCSI controller type 'lsiLogic-sas'"

### ISSUE THREAD

**Issue by KingJ** [kubernetes/kubernetes#37527]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): vsphere, lsilogic-sas

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug Report

**Kubernetes version** (use `kubectl version`): Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5+coreos.0", GitCommit:"f70c2e5b2944cb5d622621a706bdec3d8a5a9c5e", GitTreeState:"clean", BuildDate:"2016-10-31T19:16:47Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**: vSphere (vCenter: 6.5.0, ESXi: 6.0.0)
- **OS** (e.g. from /etc/os-release): CoreOS 1185.3.0 (MoreOS)
- **Kernel** (e.g. `uname -a`): Linux 4.7.3-coreos-r2 #1 SMP Tue Nov 1 01:38:43 UTC 2016 x86_64
- **Install tools**:
- **Others**:

**What happened**:
During pod provisioning, kube-controller-manager logged the following messages whilst attempting to attach a vSphere volume to the pod;

```
2016-11-27T14:15:01.636686228Z I1127 14:15:01.636561       1 reconciler.go:168] Started AttachVolume for volume "kubernetes.io/vsphere-volume/[DATASTORE] volumes/VOLUME" to node "NODE"
2016-11-27T14:15:02.085465901Z E1127 14:15:02.085346       1 vsphere.go:667] error creating new SCSI controller: unknown SCSI controller type 'lsiLogic-sas'
2016-11-27T14:15:02.102770470Z E1127 14:15:02.102592       1 attacher.go:79] Error attaching volume "[DATASTORE] volumes/VOLUME": unknown SCSI controller type 'lsiLogic-sas'
2016-11-27T14:15:02.102790243Z E1127 14:15:02.102691       1 nestedpendingoperations.go:253] Operation for "\"kubernetes.io/vsphere-volume/[DATASTORE] volumes/VOLUME\"" failed. No retries permitted until 2016-11-27 14:15:06.102667758 +0000 UTC (durationBeforeRetry 4s). Error: Failed to attach volume "VOLUME" on node "NODE" with: unknown SCSI controller type 'lsiLogic-sas'
2016-11-27T14:15:02.102908054Z I1127 14:15:02.102838       1 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"mayan-deployment-2304867309-f4tgb", UID:"df9106ba-b4ab-11e6-b3bc-005056892975", APIVersion:"v1", ResourceVersion:"4951433", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "VOLUME" on node "NODE" with: unknown SCSI controller type 'lsiLogic-sas'
```

The operation is attempted again and continues to fail. The pod becomes stuck in the ContainerCreating state.

**What you expected to happen**:
The volume should be mounted to the host and attached to the pod.

**How to reproduce it** (as minimally and precisely as possible):
1) Create a VMDK using the following command;
```
vmkfstools -c 2G /vmfs/volumes/DATASTORE/volumes/VOLUME.vmdk
```
2) Attach a vSphere volume to a pod using the following spec;
```
      volumes:
        - name: VOLUME
          vsphereVolume:
            volumePath: "[DATASTORE] volumes/VOLUME"
            fsType: ext4
```

**Anything else do we need to know**:
This happens both with attaching an existing vSphere volume (as documented [here](http://kubernetes.io/docs/user-guide/volumes/#vspherevolume)) and with the vSphere StorageClass provisioner (as documented [here](http://kubernetes.io/docs/user-guide/persistent-volumes/#vsphere)). For the purposes of documenting this issue, I followed the former rather than the latter, but the error is the same. 
Volume/datastore/host names have been substituted from their actual values.
The Kubernetes worker node's VM hardware is as follows;
![image](https://cloud.githubusercontent.com/assets/6966796/20649056/3dbc5894-b4ae-11e6-8830-51bb61b45eac.png)

**Comment by KingJ** [comment]
Just tried updating to v1.4.6 and i'm seeing the same result. @kerneltime would you have any insight into what might be causing this? (apologies for mentioning you directly!)

**Comment by jingxu97** [comment]
cc @kubernetes/sig-storage

**Comment by rootfs** [comment]
@abithap

**Comment by KingJ** [comment]
After reading over vpshere.go, I saw that there's a [configuration setting for the SCSI Controller Type](https://github.com/kubernetes/kubernetes/blob/v1.4.6/pkg/cloudprovider/providers/vsphere/vsphere.go#L110). I added the following lines to my vsphere cloud provider configuration on the K8 master and worker and restarted the kubelet daemon;

```
      [Disk]
      scsicontrollertype = pvscsi
```

Watching the kube-controller-manager log, it looks like this was the trick to getting the volume to mount;

```
2016-12-02T22:47:06.679421906Z I1202 22:47:06.679238       1 operation_executor.go:612] AttachVolume.Attach succeeded for volume "kubernetes.io/vsphere-volume/[DATASTORE]volumes/VOLUME" (spec.Name: "VOLUME") from node "NODE".
```

However, I can't find any mention of this configuration setting in the documentation. At a guess, if you don't specify a SCSI Controller Type in the cloud provider configuration, it defaults to lsiLogic-scsi, which causes the first error listed.

Although the volume is now mounted to the VM, as verified by examining the VM's hardware properties in vCenter, I am experiencing a seperate issue whereby the volume times out on attachment to the pod. This causes the pod to get stuck in the ContainerCreating state, but i'll discuss that on a separate issue.
